src/internal/pubsub-client.ts

```diff
@@ -0,0 +1,316 @@
+          if (resp.item.value.text) {
+            options.dataListener(new TopicSubscribe.Item(resp.item.value.text));
+          } else if (resp.item.value.binary) {
+            options.dataListener(
+              new TopicSubscribe.Item(resp.item.value.binary)
+            );
+          }
+        } else if (resp?.heartbeat) {
+          this.logger.trace(
+            'Received heartbeat from subscription stream; topic: %s',
+            truncateString(topicName)
+          );
+        } else if (resp?.discontinuity) {
+          this.logger.trace(
+            'Received discontinuity from subscription stream; topic: %s',
+            truncateString(topicName)
+          );
+        }
+      })
+      .on('error', (err: Error) => {
+        const serviceError = err as unknown as ServiceError;
+        // The service cuts the the stream after ~1 minute. Hence we reconnect.
+        if (
+          serviceError.code === Status.INTERNAL &&
+          serviceError.details === 'Received RST_STREAM with code 0'
+        ) {
+          this.logger.trace(
+            'Server closed stream due to idle activity. Restarting.'
+          );
+          this.sendSubscribe(
+            cacheName,
+            topicName,
+            options,
+            lastTopicSequenceNumber + 1
+          );
+          restartedDueToError = true;
+          return;
+        }
+
+        // Otherwise we propagate the error to the caller.
+        options.errorListener(
+          new TopicSubscribe.Error(cacheServiceErrorMapper(serviceError))
+        );
+      })
+      .on('end', () => {
+        // The stream could have already been restarted due to an error.
+        if (restartedDueToError) {
+          this.logger.trace(
+            'Stream ended after error but was restarted on topic: %s',
+            topicName
+          );
+          return;
+        }
+
+        this.logger.trace('Stream ended on topic: %s; restarting.', topicName);
+        this.sendSubscribe(
+          cacheName,
+          topicName,
+          options,
+          lastTopicSequenceNumber + 1
+        );
+      });
+  }
+
+  private initializeUnaryInterceptors(
+    loggerFactory: MomentoLoggerFactory,
+    middlewares: Middleware[]
+  ): Interceptor[] {
+    return this.initializeInterceptors(
+      loggerFactory,
+      middlewares,
+      this.unaryRequestTimeoutMs
+    );
+  }
+
+  private initializeStreamingInterceptors(
+    loggerFactory: MomentoLoggerFactory,
+    middlewares: Middleware[]
+  ): Interceptor[] {
+    return this.initializeInterceptors(loggerFactory, middlewares);
+  }
+
+  private initializeInterceptors(
+    loggerFactory: MomentoLoggerFactory,
+    middlewares: Middleware[],
+    timeoutMs?: number
+  ): Interceptor[] {
+    const headers = [
+      new Header('Authorization', this.credentialProvider.getAuthToken()),
+      new Header('Agent', `nodejs:${version}`),
+    ];
+    if (timeoutMs) {
```

Comment:
I don't really like keying off of this.  We should just implement the streaming/unary interceptors functions separately for now.  And you should remove the loggerFactory/middleWare args from the signature for the Streaming one for now.  That will unblock us to merge this and then come back and revisit.

Reply:
Refactored in [2647f36](https://github.com/momentohq/client-sdk-nodejs/pull/347/commits/2647f36a2e5508451fb7001f08dc93a013196076)